File: leetcode-subscription/c#/Problems/P0253.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace LeetCode.Naive.Problems
{
  /// <summary>
  ///    Problem: https://leetcode.com/problems/meeting-rooms-ii/
  ///    Submission: https://leetcode.com/submissions/detail/451187721/
  ///    All
  /// </summary>
  internal class P0253
  {
    public class Solution
    {
      public int MinMeetingRooms(int[][] intervals)
      {
        var items = intervals
          .SelectMany(i => new[] { (value: i[0], dir: 1), (value: i[1], dir: 0) })
          .OrderBy(i => i.value)
          .ThenBy(i => i.dir)
          .ToArray();

        var ans = 0;
        var current = 0;

        foreach (var item in items)
        {
          if (item.dir == 1)
            current++;
          else
            current--;

          ans = Math.Max(ans, current);
        }

        return ans;
      }
    }
  }
}

File: binarysearch/src/main/java/binarysearch/P0253.java
package binarysearch;

/*
 * Problem: https://binarysearch.com/problems/Number-of-Hops
 * Submission: https://binarysearch.com/problems/Number-of-Hops/submissions/7389502
 */
public class P0253 {
  class Solution {
    public int solve(int[] nums) {
      var n = nums.length;
      if (n <= 1)
        return 0;

      var maxEnd = nums[0];
      var count = 1;

      if (maxEnd >= n - 1)
        return count;

      for (var i = 0;; ) {

        var currentEnd = maxEnd;

        while (i <= currentEnd) {
          maxEnd = Math.max(i + nums[i], maxEnd);
          if (maxEnd >= n - 1)
            return count + 1;

          i++;
        }

        count++;
      }
    }
  }
}